File: assignment_7/service.c
#include <stdio.h>
#include <stdlib.h>

int compare(const void *a, const void *b) {
    return (*(int*)a - *(int*)b);
}

double mean(int arr[], int n) {
    int sum = 0;
    for (int i = 0; i < n; i++) {
        sum += arr[i];
    }
    return (double)sum / n;
}

double median(int arr[], int n) {
    qsort(arr, n, sizeof(int), compare);
    if (n % 2 == 0)
        return (arr[n/2 - 1] + arr[n/2]) / 2.0;
    else
        return arr[n/2];
}

void mode(int arr[], int n, int result[], int *mode_count) {
    int max_count = 0;
    *mode_count = 0;
    for (int i = 0; i < n; i++) {
        int count = 1;
        for (int j = i + 1; j < n; j++) {
            if (arr[j] == arr[i]) count++;
        }
        if (count > max_count) {
            max_count = count;
            result[0] = arr[i];
            *mode_count = 1;
        } else if (count == max_count) {
            int found = 0;
            for (int k = 0; k < *mode_count; k++) {
                if (result[k] == arr[i]) {
                    found = 1;
                    break;
                }
            }
            if (!found) {
                result[*mode_count] = arr[i];
                (*mode_count)++;
            }
        }
    }
}

int main() {
    int arr[] = {1, 2, 2, 3, 4};
    int n = sizeof(arr)/sizeof(arr[0]);

    printf("Mean: %.2f\n", mean(arr, n));
    printf("Median: %.2f\n", median(arr, n));

    int mode_result[100];
    int mode_count;
    mode(arr, n, mode_result, &mode_count);
    printf("Mode: ");
    for (int i = 0; i < mode_count; i++) {
        printf("%d ", mode_result[i]);
    }
    printf("\n");
    return 0;
}

File: assignment_7/service.ml
let mean lst =
  let total = List.fold_left ( +. ) 0.0 (List.map float_of_int lst) in
  total /. float_of_int (List.length lst)

let median lst =
  let sorted = List.sort compare lst in
  let len = List.length sorted in
  if len mod 2 = 0 then
    let mid1 = List.nth sorted (len / 2 - 1) in
    let mid2 = List.nth sorted (len / 2) in
    float_of_int (mid1 + mid2) /. 2.0
  else
    float_of_int (List.nth sorted (len / 2))

let mode lst =
  let tbl = Hashtbl.create 10 in
  List.iter (fun x ->
    Hashtbl.replace tbl x ((try Hashtbl.find tbl x with Not_found -> 0) + 1)) lst;
  let max_count = Hashtbl.fold (fun _ count acc -> max count acc) tbl 0 in
  Hashtbl.fold (fun k v acc -> if v = max_count then k :: acc else acc) tbl []

let () =
  let data = [1; 2; 2; 3; 4] in
  Printf.printf "Mean: %.2f\n" (mean data);
  Printf.printf "Median: %.2f\n" (median data);
  Printf.printf "Mode: %s\n"
    (String.concat " " (List.map string_of_int (mode data)))
